### PR TITLE
Enable Overwatch for Unity Catalog

### LINKED ISSUE #476: [EPIC] - Enable OW on UC
Enable OW on UC

- [x] OW should be able to run on UC as a first party dataset
- [x] migration path must be created for users to migrate

Note internal slack convo here:
https://databricks.slack.com/archives/C012VT7C136/p1659384126896419

So far it looks like UC can work but may need some specific docs added to call this out.

### PR DESCRIPTION
**Pull request by aman-db**
closes #476

### CODE REVIEW

File: src/main/scala/com/databricks/labs/overwatch/pipeline/InitializerFunctionsUCE.scala [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+    val dbLocation = PipelineFunctions.cleansePathURI(rawDBLocation)
+    val rawETLDataLocation = dataTarget.etlDataPathPrefix.getOrElse(dbLocation)
+    val etlDataLocation = PipelineFunctions.cleansePathURI(rawETLDataLocation)
+    val etlCatalogName = dataTarget.databaseName.get.split("\\.").head
+    var switch = true
+
+
+    spark.sessionState.catalogManager.setCurrentCatalog(etlCatalogName)
+    if (spark.catalog.databaseExists(dbName)) {
+      val dbMeta = spark.sessionState.catalog.getDatabaseMetadata(dbName)
```
Comment (GeekSheikh) [inline src/main/scala/com/databricks/labs/overwatch/pipeline/InitializerFunctionsUCE.scala:35]: all this metadata stuff still works with UC? If so, great, just wanted to make sure you double checked.

File: src/main/scala/com/databricks/labs/overwatch/pipeline/InitializerFunctionsUCE.scala [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+  }
+    /**
+     * validate and set the DataTarget for Overwatch
+     * @param dataTarget OW DataTarget
+     */
+    override def validateAndSetDataTarget(dataTarget: DataTarget): Unit = {
+      // Validate data Target
+      // todo UC enablement
+
+      val etlCatalogName = dataTarget.databaseName.get.split("\\.").head
```
Comment (GeekSheikh) [inline src/main/scala/com/databricks/labs/overwatch/pipeline/InitializerFunctionsUCE.scala:136]: I believe you've already verified that the databaseName has a "." in it and that's how we got to this point in the workflow in the first place. I just get nervous anytime I see .head as it will blow up the entire pipeline if the value is empty. Please verify that we KNOW split(...).head will return a value -- if you have any doubt put an appropriate error handler her and throw a user-friendly message.

File: src/main/scala/com/databricks/labs/overwatch/pipeline/InitializerFunctionsUCE.scala [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+      spark.sessionState.catalogManager.setCurrentCatalog(etlCatalogName)
+      if (!disableValidations) dataTargetIsValid(dataTarget)
+
+      // If data target is valid get db name and location and set it
+      val dbName = dataTarget.databaseName.get.split("\\.").last
```
Comment (GeekSheikh) [inline src/main/scala/com/databricks/labs/overwatch/pipeline/InitializerFunctionsUCE.scala:141]: same here -- @sriram251-code / you may need to add some logic to MW config validation to ensure users don't try to use more than one period if we're going to rely on it so heavily. What if user does something like catalog.oops.schema -- do we identify that this is an error?

File: src/main/scala/com/databricks/labs/overwatch/pipeline/Module.scala [hunk 4]
```diff
@@ -351,6 +351,7 @@ class Module(
   @throws(classOf[IllegalArgumentException])
   def execute(_etlDefinition: () => ETLDefinition): ModuleStatusReport = {
 //    _etlDefinition: ETLDefinition
+    spark.sessionState.catalogManager.setCurrentCatalog(config.etlCatalogName)
```
Comment (GeekSheikh) [inline src/main/scala/com/databricks/labs/overwatch/pipeline/Module.scala:354]: Since you set this in the Initializer, why do you keep setting it throughout the code?

File: src/main/scala/com/databricks/labs/overwatch/utils/Tools.scala [hunk 5]
```diff
@@ -182,6 +182,23 @@ object Helpers extends SparkSessionWrapper {
...
+   */
+  def pathPatternExists(name: String): Boolean = {
+    val path = new Path(name)
+    val fs = path.getFileSystem(spark.sparkContext.hadoopConfiguration)
+    val globPath = fs.globStatus(path)
+    if (globPath.isEmpty)
+      false
+    else
+       true
+  }
```
Comment (GeekSheikh) [inline src/main/scala/com/databricks/labs/overwatch/utils/Tools.scala:199]: This is still very strange to me -- I think it's what we discussed but please double think just to make sure .

Also -- please make this private[overwatch] -- no user should be using this.

File: src/main/scala/com/databricks/labs/overwatch/utils/Tools.scala [hunk 6]
```diff
@@ -560,13 +577,20 @@ object Helpers extends SparkSessionWrapper {
    * @return
    */
   def getWorkspaceByDatabase(
-                              etlDB: String,
+//                              etlDB: String,
+                              db: String,
```
Comment (GeekSheikh) [inline src/main/scala/com/databricks/labs/overwatch/utils/Tools.scala:581]: why did you change the name? it was etlDB so users no that this cannot reference the consumer DB

File: src/main/scala/com/databricks/labs/overwatch/utils/Tools.scala [hunk 7]
```diff
@@ -560,13 +577,20 @@ object Helpers extends SparkSessionWrapper {
...
                               organization_id: Option[String] = None,
                               apiUrl: Option[String] = None,
                               successfullOnly: Boolean = true,
                               disableValidations: Boolean = false
                             ): Workspace = {
     // verify database exists
+    val initialCatalog = spark.sessionState.catalogManager.currentCatalog.name()
+    val etlDB = if(db.contains(".")){
+      spark.sessionState.catalogManager.setCurrentCatalog(db.split("\\.").head)
```
Comment (GeekSheikh) [inline src/main/scala/com/databricks/labs/overwatch/utils/Tools.scala:590]: This code is very redundant throughout -- please create a Helper function like

setCatalog(spark, databaseName) = {...} and use this anywhere you need to use this. Imperative that you pass in the spark session though to ensure thread-local settings.

Comment (GeekSheikh) [inline src/main/scala/com/databricks/labs/overwatch/utils/Tools.scala:590]: is this necessary? If user passes in "catalog.schema" wouldn't the spark.table call just read spark.table("catalog.schema.pipeline_report") which would work anyway? Asking, not telling here. Just curious

File: src/main/scala/com/databricks/labs/overwatch/validation/Kitana.scala [hunk 8]
```diff
@@ -367,6 +367,7 @@ object Kitana {
       databaseName = Some(snapDBName),
       databaseLocation = snapDBLocation,
       etlDataPathPrefix = snapDBDataLocation,
+//      catalogName =  Some("overwatch"),
```
Comment (GeekSheikh) [inline src/main/scala/com/databricks/labs/overwatch/validation/Kitana.scala:370]: remove this commented line if not needed.

File: src/test/scala/com/databricks/labs/overwatch/ParamDeserializerTest.scala [hunk 9]
```diff
@@ -52,6 +52,7 @@ class ParamDeserializerTest extends AnyFunSpec {
         externalizeOptimize = false,
         apiEnvConfig = None,
         tempWorkingDir = ""
+//        deploymentType = "default"
```
Comment (GeekSheikh) [inline src/test/scala/com/databricks/labs/overwatch/ParamDeserializerTest.scala:55]: remove this commented line if not needed.

File: src/test/scala/com/databricks/labs/overwatch/pipeline/InitializerFunctionsTest.scala [hunk 10]
```diff
@@ -2,11 +2,13 @@ package com.databricks.labs.overwatch.pipeline
 import com.databricks.labs.overwatch.SparkSessionTestWrapper
 import com.databricks.labs.overwatch.utils.Config
 import com.github.mrpowers.spark.fast.tests.DataFrameComparer
+import org.scalatest.{Ignore, color}
 import org.scalatest.funspec.AnyFunSpec
-
+//@Ignore
```
Comment (GeekSheikh) [inline src/test/scala/com/databricks/labs/overwatch/pipeline/InitializerFunctionsTest.scala:7]: delete this commented ignore line

File: src/test/scala/com/databricks/labs/overwatch/pipeline/InitializeTest.scala [hunk 11]
```diff
@@ -7,18 +7,20 @@ import com.databricks.labs.overwatch.utils.{BadConfigException, Config, Overwatc
...
     it ("should validate isPVC as false when org id if doesn't have ilb") {
       val conf = new Config
       conf.setOrganizationId("demo")
+      conf.setDeploymentType("default")
       val init = new Initializer(conf, disableValidations = false, isSnap = false, initDB = true)
-
-      val isPVC = PrivateMethod[Boolean]('isPVC)
-      val actual = init.Init invokePrivate isPVC()
+//      val isPVC = PrivateMethod[Boolean]('isPVC)
+//      val actual = init.Init invokePrivate isPVC()
```
Comment (GeekSheikh) [inline src/test/scala/com/databricks/labs/overwatch/pipeline/InitializeTest.scala:22]: remove this commented line if not needed.

Commented out lines have been left throughout this file -- please remove all your commented lines that are no longer needed in this file.

**Review by GeekSheikh — CHANGES_REQUESTED** [review]
Looks good -- minor changes requested -- mostly cosmetic.

File: src/main/scala/com/databricks/labs/overwatch/pipeline/Initializer.scala [hunk 12]
```diff
@@ -16,42 +16,24 @@ class Initializer(config: Config, disableValidations: Boolean, isSnap: Boolean,
 
   private val logger: Logger = Logger.getLogger(this.getClass)
 
+  val Init = InitializerFunctions(config, disableValidations, isSnap, initDB)
```
[the remark this answers is not in the document]
Comment (aman-db) [reply src/main/scala/com/databricks/labs/overwatch/pipeline/Initializer.scala:19]: @GeekSheikh - we are creating Init to have either Default InitializerFunctions or UCE InitializerFunctions which will be decided by `config.deploymentType`

File: src/main/scala/com/databricks/labs/overwatch/pipeline/Initializer.scala [hunk 13]
```diff
@@ -16,42 +16,24 @@ class Initializer(config: Config, disableValidations: Boolean, isSnap: Boolean,
...
-     * when coming from Datbaricks jobs ui since it has to be escaped to go through the api calls.
-     *
-     * Otherwise -- read from the args passed in and serialize into OverwatchParams
-     */
-    logger.log(Level.INFO, "Validating Input Parameters")
-    val rawParams = mapper.readValue[OverwatchParams](overwatchArgs)
-
+  private def validateAndRegisterArgs(overwatchArgs: OverwatchParams): this.type = {
 
+    val rawParams = overwatchArgs
```
[the remark this answers is not in the document]
Comment (aman-db) [reply src/main/scala/com/databricks/labs/overwatch/pipeline/Initializer.scala:33]: changed the name to rawParams in function signature

File: src/main/scala/com/databricks/labs/overwatch/pipeline/InitializerFunctions.scala [hunk 14]
```diff
@@ -262,18 +262,18 @@ abstract class InitializerFunctions(config: Config, disableValidations: Boolean,
...
+//      val dbName = dataTarget.databaseName.get
+//      val dbLocation = dataTarget.databaseLocation.getOrElse(s"dbfs:/user/hive/warehouse/${dbName}.db")
+//      val dataLocation = dataTarget.etlDataPathPrefix.getOrElse(dbLocation)
+//
+//      val consumerDBName = dataTarget.consumerDatabaseName.getOrElse(dbName)
+//      val consumerDBLocation = dataTarget.consumerDatabaseLocation.getOrElse(s"/user/hive/warehouse/${consumerDBName}.db")
+//
+//      config.setDatabaseNameAndLoc(dbName, dbLocation, dataLocation)
+//      config.setConsumerDatabaseNameandLoc(consumerDBName, consumerDBLocation)
     }
```
[the remark this answers is not in the document]
Comment (aman-db) [reply src/main/scala/com/databricks/labs/overwatch/pipeline/InitializerFunctions.scala:277]: done

File: src/test/scala/com/databricks/labs/overwatch/pipeline/InitializerFunctionsTest.scala (hunk 10 above)
Comment (aman-db) [reply src/test/scala/com/databricks/labs/overwatch/pipeline/InitializerFunctionsTest.scala:7]: deleted

File: src/main/scala/com/databricks/labs/overwatch/validation/Kitana.scala (hunk 8 above)
Comment (aman-db) [reply src/main/scala/com/databricks/labs/overwatch/validation/Kitana.scala:370]: removed

File: src/main/scala/com/databricks/labs/overwatch/utils/Tools.scala (hunk 7 above)
Comment (aman-db) [reply src/main/scala/com/databricks/labs/overwatch/utils/Tools.scala:590]: @GeekSheikh  spark.sessionState.catalog.getDatabaseMetadata(dbName) - this function does not work with catalog.dbName , it fails if input is passed in this format. Thats why we have do add this logic here

File: src/main/scala/com/databricks/labs/overwatch/utils/Tools.scala (hunk 6 above)
Comment (aman-db) [reply src/main/scala/com/databricks/labs/overwatch/utils/Tools.scala:581]: I have derived etlDB through db params, so that I dont have to change this variable name through out the function. can I change db to etlDatabaseName ?

File: src/main/scala/com/databricks/labs/overwatch/pipeline/Module.scala (hunk 4 above)
Comment (aman-db) [reply src/main/scala/com/databricks/labs/overwatch/pipeline/Module.scala:354]: After each layer completion it will set to default hive_metastore

File: src/main/scala/com/databricks/labs/overwatch/pipeline/InitializerFunctionsUCE.scala [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+    val consumerDBLocation = PipelineFunctions.cleansePathURI(rawConsumerDBLocation)
+    val consumerCatalogName = dataTarget.consumerDatabaseName.getOrElse(dbName).split("\\.").head
+    spark.sessionState.catalogManager.setCurrentCatalog(consumerCatalogName)
+    if (consumerDBName != dbName) { // separate consumer db
+      if (spark.catalog.databaseExists(consumerDBName)) {
+        val consumerDBMeta = spark.sessionState.catalog.getDatabaseMetadata(consumerDBName)
+
+        val existingConsumerDBLocation = spark.sql(s"describe database ${consumerDBName}")
+          .where("database_description_item ='RootLocation'")
+          .select("database_description_value").collect.map(_.getString(0)).head.toString
```
[the remark this answers is not in the document]
Comment (aman-db) [reply src/main/scala/com/databricks/labs/overwatch/pipeline/InitializerFunctionsUCE.scala:93]: moved line 82 to 85 above (around 24)

File: src/main/scala/com/databricks/labs/overwatch/pipeline/InitializerFunctionsUCE.scala (hunk 1 above)
Comment (aman-db) [reply src/main/scala/com/databricks/labs/overwatch/pipeline/InitializerFunctionsUCE.scala:35]: yes it does work with UC

File: src/main/scala/com/databricks/labs/overwatch/pipeline/Initializer.scala (hunk 12 above)
[the remark this answers is not in the document]
Comment (GeekSheikh) [reply src/main/scala/com/databricks/labs/overwatch/pipeline/Initializer.scala:19]: I feel like this should be set in the apply method and the constructor should already have the proper deployment type. Let's discuss further.

File: src/main/scala/com/databricks/labs/overwatch/pipeline/BronzeTransforms.scala [hunk 16]
```diff
@@ -637,8 +637,7 @@ trait BronzeTransforms extends SparkSessionWrapper {
         .filter(!'failed && 'withinSpecifiedTimeRange)
         .select('filename)
         .distinct
-
-      if (Helpers.pathExists(badRecordsPath)) {
+      if (Helpers.pathExists(badRecordsPath) && Helpers.pathPatternExists(s"${badRecordsPath}/*/*/")) {
```
[the remark this answers is not in the document]
Comment (GeekSheikh) [reply src/main/scala/com/databricks/labs/overwatch/pipeline/BronzeTransforms.scala:640]: let's clarify in discussion

[the remark this answers is not in the document]
Comment (GeekSheikh) [reply src/main/scala/com/databricks/labs/overwatch/pipeline/BronzeTransforms.scala:640]: understood -- add comments to clarify above the if and a little more above the pathPatternExists func

File: src/main/scala/com/databricks/labs/overwatch/pipeline/InitializerFunctions.scala (hunk 14 above)
[the remark this answers is not in the document]
Comment (GeekSheikh) [reply src/main/scala/com/databricks/labs/overwatch/pipeline/InitializerFunctions.scala]: update interface definition as per industry standard

File: src/main/scala/com/databricks/labs/overwatch/pipeline/InitializerFunctions.scala [hunk 17]
```diff
@@ -428,38 +428,6 @@ abstract class InitializerFunctions(config: Config, disableValidations: Boolean,
...
-          val createConsumerDBSTMT = s"create database if not exists ${config.consumerDatabaseName} " +
-            s"location '${config.consumerDatabaseLocation}'"
-
-          spark.sql(createConsumerDBSTMT)
-          logger.log(Level.INFO, s"Successfully created database. $createConsumerDBSTMT")
-        }
-      }
-
       Database(config)
     }
```
[the remark this answers is not in the document]
Comment (GeekSheikh) [reply src/main/scala/com/databricks/labs/overwatch/pipeline/InitializerFunctions.scala:419]: Refactor and simplify -- only override the components that are actually different not the entire function

File: src/main/scala/com/databricks/labs/overwatch/pipeline/InitializerFunctionsUCE.scala [hunk 18]
```diff
@@ -0,0 +1,210 @@
...
+      // Validate data Target
+      // todo UC enablement
+
+      val etlCatalogName = dataTarget.databaseName.get.split("\\.").head
+      spark.sessionState.catalogManager.setCurrentCatalog(etlCatalogName)
+      if (!disableValidations) dataTargetIsValid(dataTarget)
+
+      // If data target is valid get db name and location and set it
+      val dbName = dataTarget.databaseName.get.split("\\.").last
+      val dbLocation = dataTarget.databaseLocation.getOrElse(s"dbfs:/user/hive/warehouse/${dbName}.db")
```
[the remark this answers is not in the document]
Comment (GeekSheikh) [reply src/main/scala/com/databricks/labs/overwatch/pipeline/InitializerFunctionsUCE.scala:142]: dbLoc and consumerDBLocs are not needed for UC as customers will be creating the schemas and provisioning access; thus it doesn't matter where the databases are since all the tables are external

File: src/main/scala/com/databricks/labs/overwatch/pipeline/InitializerFunctionsUCE.scala (hunk 3 above)
Comment (GeekSheikh) [reply src/main/scala/com/databricks/labs/overwatch/pipeline/InitializerFunctionsUCE.scala:141]: Use lastOption and add this to the DataTarget funcs

File: src/main/scala/com/databricks/labs/overwatch/pipeline/BronzeTransforms.scala (hunk 16 above)
[the remark this answers is not in the document]
Comment (aman-db) [reply src/main/scala/com/databricks/labs/overwatch/pipeline/BronzeTransforms.scala:640]: done

File: src/main/scala/com/databricks/labs/overwatch/pipeline/InitializerFunctionsUCE.scala [hunk 19]
```diff
@@ -0,0 +1,210 @@
...
+
+
+    spark.sessionState.catalogManager.setCurrentCatalog(etlCatalogName)
+    if (spark.catalog.databaseExists(dbName)) {
+      val dbMeta = spark.sessionState.catalog.getDatabaseMetadata(dbName)
+      val dbProperties = dbMeta.properties
+
+      val existingDBLocation = spark.sql(s"describe database ${dbName}")
+        .where("database_description_item ='RootLocation'")
+        .select("database_description_value").collect.map(_.getString(0)).head.toString
```
[the remark this answers is not in the document]
Comment (aman-db) [reply src/main/scala/com/databricks/labs/overwatch/pipeline/InitializerFunctionsUCE.scala:44]: done

File: src/main/scala/com/databricks/labs/overwatch/pipeline/InitializerFunctionsUCE.scala (hunk 3 above)
Comment (aman-db) [reply src/main/scala/com/databricks/labs/overwatch/pipeline/InitializerFunctionsUCE.scala:141]: done

File: src/main/scala/com/databricks/labs/overwatch/utils/Tools.scala (hunk 6 above)
Comment (aman-db) [reply src/main/scala/com/databricks/labs/overwatch/utils/Tools.scala:581]: done

File: src/main/scala/com/databricks/labs/overwatch/utils/Tools.scala (hunk 7 above)
Comment (aman-db) [reply src/main/scala/com/databricks/labs/overwatch/utils/Tools.scala:590]: done. created two different functions to get and set the catalog

File: src/main/scala/com/databricks/labs/overwatch/pipeline/Initializer.scala (hunk 12 above)
[the remark this answers is not in the document]
Comment (aman-db) [reply src/main/scala/com/databricks/labs/overwatch/pipeline/Initializer.scala]: done

File: src/main/scala/com/databricks/labs/overwatch/pipeline/InitializerFunctions.scala (hunk 17 above)
[the remark this answers is not in the document]
Comment (aman-db) [reply src/main/scala/com/databricks/labs/overwatch/pipeline/InitializerFunctions.scala]: done. Removed the db location checks for UCE

File: src/main/scala/com/databricks/labs/overwatch/utils/Tools.scala (hunk 7 above)
Comment (aman-db) [reply src/main/scala/com/databricks/labs/overwatch/utils/Tools.scala]: done.

File: src/main/scala/com/databricks/labs/overwatch/pipeline/Initializer.scala [hunk 20]
```diff
@@ -123,7 +123,7 @@ object Initializer extends SparkSessionWrapper {
                         disableValidations: Boolean = false,
                         isSnap: Boolean = false,
                         initializeDatabase: Boolean = true): Initializer = {
-    config.deploymentType match {
+    config.deploymentType.toLowerCase.trim match {
```
Comment (GeekSheikh) [inline src/main/scala/com/databricks/labs/overwatch/pipeline/Initializer.scala:126]: good catch

File: src/main/scala/com/databricks/labs/overwatch/MultiWorkspaceDeployment.scala [hunk 21]
```diff
@@ -427,6 +427,7 @@ class MultiWorkspaceDeployment extends SparkSessionWrapper {
     Future {
       val threadDeploymentReport = ArrayBuffer[MultiWSDeploymentReport]()
       val deploymentId = deploymentParams.deploymentId
+      println(s"deploymentParams.args.toString-------${deploymentParams.args}")
```
Comment (GeekSheikh) [inline src/main/scala/com/databricks/labs/overwatch/MultiWorkspaceDeployment.scala:430]: was this for debug? did you mean to leave it?

**Review by GeekSheikh — CHANGES_REQUESTED** [review]
lookin good

**Review by GeekSheikh — APPROVED** [review]
Congrats -- thanks for sticking with it!
